### PR TITLE
Fully document #193 stream module

### PR DESCRIPTION
**Pull request by noelmcloughlin**
This PR documents tricky nginx configuration discussed in #193.

The pillar.example is based on [this example](https://docs.nginx.com/nginx/admin-guide/load-balancer/tcp-udp-load-balancer/#example ) and renders config shown.  Sorry for the ugly yaml.
```
stream {
    server {
        listen 1000;
        proxy_pass lb-1000;
    }
    server  {
        listen 53 udp;
        proxy_pass dns_servers;
    }
    server   {
        listen 12346;
        proxy_pass backend4.example.com:12346;
    }
    upstream dns_servers {

        server 192.168.136.130:53 ;
        server 192.168.136.131:53 ;
        server 192.168.136.132:53 ;
    }

    upstream lb-1000 {

        server hostname1.example.com:1000;
        server hostname2.example.com:1000;
    }
    upstream stream_backend {
        least_conn ;
        server backend1.example.com:12345 weight=5;
        server backend2.example.com:12345 max_fails=2 fail_timeout=30s ;
        server backend3.example.com:12345 max_conns=3 ;
    }
}
user www-data;
worker_processes 4;
```

### CODE REVIEW

File: pillar.example [hunk 1]
```diff
@@ -119,6 +120,26 @@ nginx:
...
             proxy_pass: lb-1000
 
+          upstream stream_backend:
+            least_conn: ''
+            'server': 'backend1.example.com:12345 weight=5'                       #one way
+            'server backend2.example.com:12345 max_fails=2 fail_timeout=30s': ''   
+            'server backend3.example.com:12345 max_conns=3': ''                   #another way
+          upstream dns_servers:
+            least_conn:
+            'server 192.168.136.130:53': ''
```
Comment (noelmcloughlin) [inline pillar.example]: The '' is maybe not needed?

**Comment by noelmcloughlin** [comment]
It might be possible to simplify few bits - suggestions welcome!!

**Comment by aboe76** [comment]
@noelmcloughlin I use snippets for this stuff:

```
nginx:
  ng:
    snippets:
      upstream_netdata_tcp:
        - upstream netdata:
          - server: 127.0.0.1:19999
          - keepalive: 64
```

**Comment by noelmcloughlin** [comment]
Questions:
Can you model `least_conn ;` in snippets?
Can you have duplicate keys?
Is snippets better solution for pillar.example overall?

We should ensure the pillar.example have enough examples that yaml is not blocking them.

**Comment by aboe76** [comment]
@noelmcloughlin 
```
nginx:                                                                                                       [0/604]
  ng:
    snippets:
      upstream_netdata_tcp:
        - upstream netdata:
          - server: 127.0.0.1:19999
          - keepalive: 64
      upstream_phpcgi_unix:
        - upstream phpcgi:
          - least_conn: ''
          - server: unix:/var/run/php-fpm/php-fpm-arch.sock                                                        
          - server: unix:/var/run/php-fpm/php-fpm-arch-two.sock
```

**Comment by noelmcloughlin** [comment]
Excellent ... I'll update the PR later. thanks!!!

**Comment by aboe76** [comment]
@noelmcloughlin snippets are a way to split nginx configs in smaller more managable items,
which can be included in other files.

I have used this formula to mimic:  https://github.com/perusio/drupal-with-nginx

**Comment by aboe76** [comment]
@noelmcloughlin I don't know, I use pillarstack and varstack to combine my pillar data...

**Comment by aboe76** [comment]
But I `nginx:ng` can use multiple keys: 

see the nginx.conf for `listen`

**Comment by aboe76** [comment]
@noelmcloughlin one other thing usefull for snippets part with multiple keys:

```
nginx:
  ng:
    snippets:
      cloudflare_proxy:
        - set_real_ip_from: 103.21.244.0/22
        - set_real_ip_from: 103.22.200.0/22
        - set_real_ip_from: 103.31.4.0/22
        - set_real_ip_from: 104.16.0.0/12
        - set_real_ip_from: 108.162.192.0/18
        - set_real_ip_from: 131.0.72.0/22
        - set_real_ip_from: 141.101.64.0/18
        - set_real_ip_from: 162.158.0.0/15
        - set_real_ip_from: 172.64.0.0/13
        - set_real_ip_from: 173.245.48.0/20
        - set_real_ip_from: 188.114.96.0/20
        - set_real_ip_from: 190.93.240.0/20
        - set_real_ip_from: 197.234.240.0/22
        - set_real_ip_from: 198.41.128.0/17
        - set_real_ip_from: 2400:cb00::/32
        - set_real_ip_from: 2606:4700::/32
        - set_real_ip_from: 2803:f800::/32
        - set_real_ip_from: 2405:b500::/32
        - set_real_ip_from: 2405:8100::/32
        - set_real_ip_from: 2c0f:f248::/32
        - set_real_ip_from: 2a06:98c0::/29
```

**Comment by aboe76** [comment]
Or something truly awfull:
```
ngingx:
  ng:
    snippets:
      blacklist:
        - map $http_user_agent $bad_bot:
          - default: 0
          - '~*^Lynx': 0
          - '~*malicious': 1
          - '~*bot': 1
          - '~*backdoor': 1
          - '~*crawler': 1
          - '~*bandit': 1
          - libwww-perl: 1
          - '~(?i)(httrack|htmlparser|libwww)': 1
```

**Comment by noelmcloughlin** [comment]
Hi @aboe76 please review this now....

**Comment by aboe76** [comment]
@noelmcloughlin nice work

**Pull request merged**